Sources/NIOHTTP2/StreamMap.swift

```diff
@@ -0,0 +1,273 @@
+    // This function exists as a performance optimisation: we can keep track of the index and where we are, and
+    // thereby avoid searching the array twice.
+    //
+    /// Transform the value of an HTTP2StreamStateMachine, closing it if it ends up closed.
+    ///
+    /// - parameters:
+    ///     - modifier: A block that will modify the contained value in the
+    ///         optional, if there is one present.
+    /// - returns: The return value of the block or `nil` if the optional was `nil`.
+    mutating func autoClosingTransform<T>(streamID: HTTP2StreamID, _ modifier: (inout Element) -> T) -> T? {
+        if streamID.mayBeInitiatedBy(.client) {
+            return self.clientInitiated.autoClosingTransform(streamID: streamID, modifier)
+        } else {
+            return self.serverInitiated.autoClosingTransform(streamID: streamID, modifier)
+        }
+    }
+
+
+    // This function exists as a performance optimisation: we can keep track of the index and where we are, and
+    // thereby avoid searching the array twice.
+    mutating func transformOrCreateAutoClose<T>(streamID: HTTP2StreamID,
+                                                _ creator: () throws -> Element,
+                                                _ transformer: (inout Element) -> T) rethrows -> T? {
+        if streamID.mayBeInitiatedBy(.client) {
+            return try self.clientInitiated.transformOrCreateAutoClose(streamID: streamID, creator, transformer)
+        } else {
+            return try self.serverInitiated.transformOrCreateAutoClose(streamID: streamID, creator, transformer)
+        }
+    }
+}
+
+extension CircularBuffer where Element: PerStreamData {
+    fileprivate mutating func mutatingForEach(_ body: (inout Element) throws -> Void) rethrows {
+        var index = self.startIndex
+        let endIndex = self.endIndex
+        while index != endIndex {
+            try self.modify(index, body)
+            self.formIndex(after: &index)
+        }
+    }
+
+    fileprivate func findIndexForStreamID(_ streamID: HTTP2StreamID) -> Index? {
+        if self.count < binarySearchThreshold {
+            return self.linearScanForStreamID(streamID)
+        } else {
+            return self.binarySearchForStreamID(streamID)
+        }
+    }
+
+    private func linearScanForStreamID(_ streamID: HTTP2StreamID) -> Index? {
+        var index = self.startIndex
+
+        while index < self.endIndex {
+            let currentStreamID = self[index].streamID
+            if currentStreamID == streamID {
+                return index
+            }
+            if currentStreamID > streamID {
+                // Stop looping, we won't find this.
+                return nil
+            }
+            self.formIndex(after: &index)
+        }
+
+        return nil
+    }
+
+    // Binary search is somewhat complex code compared to a linear scan, so we don't want to inline this code if we can avoid it.
+    @inline(never)
```

Comment:
I would have just let the compiler do whatever it wants until we discover a problem.

Reply:
In general I agree. Here, because we're explicitly writing fast-path/slow-path code, I think the choice is justifiable. If we're going to do a binary search we know we have at least 200 elements to search, and so the overhead of jumping to the new method (potentially missing cache) is acceptable.

For the linear scan we may have zero or a small number of data objects. In those cases, it'd be beneficial if we inlined the relatively small amount of code necessary to run the linear search into the caller, improving locality.